Set sram0 to represent the physical RAM in nRF5340 and nRF9160 DKs

ioannisg opened the pull request:
**We would like to have `sram0 `DT node represent the available SRAM memory on nrf9160 and nRF5340 SoCs. As such, sram0 shall not be resized in Board DT configuration.**

In DTS we define sram0 node as such, but in boards//nrf5340pdk_ and nrf9160dk_ we have been using sram0 to represent (also) the image RAM allocated to Zephyr builds. The consequence is that sram0 is _resized_ to match the requirements of the specific build (security domain, inter-processor shared memory, etc.), so we do not have a DTS generated macro to represent the total SRAM.

In this PR we fix this problem by having sram0 always represent the physical SRAM. 
For nRF5340 we introduce sram0_image, to represent the RAM available for Zephyr image(s), secure and non-secure.

This PR is also adding additional information regarding reserved memory areas in board DT, such as SRAM memory areas explicitly allocated to secure and non-secure Zephyr execution environment.

hakonfam committed:
boards: arm: nRF9160: avoid resizing sram0 for nRF9160_dk

sram0 node is needed to hold the size of the
total, physical SRAM available on nRF9160 SoC.
We use sram0_s to represent the Secure image
SRAM for nRF9160_dk builds.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>
Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340pdk: define sram0_image for image(s) RAM

We should not be using sram0 for image SRAM in nrf5340pdk.
sram0 represents the physical SRAM and that one includes the
shared memory between the two M33 CPUs on the SoC. We should
not be re-sizing sram0 to account for the shared RAM; instead
we would like to have sram0 representing the whole available
SRAM.

For that, we define a new memory node, sram0_image to
represent the 'image' SRAM that is available for Zephyr
on the board. sram0_image is the chosen image SRAM for
default builds, i.e. when TrustZone is ignored
(TRUSTED_EXECUTION_SECURE is not defined).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf5340pdk: clean up sram0_shared definition

We do not want sram0_shared to represent physical ram;
this is just a portion of sram reserved for shared memory
between Application and Network MCU. Therfore, we remove
the 'mmio' compatible property and transform this node to
a reserved-memory node definition, inside which we define
the sram0_shared node along with its reg property.

In addition we correct the documentation about the shared
memory, stressing that it is placed after the image RAM of
nrf5340 Application MCU (not after the secure SRAM).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
boards: arm: nrf9160_dk: clean up sram0_ns and sram0_bsd definitions

We do not want sram0_ns and sram0_bsd to represent physical
ram; these are just portions of sram reserved for the non-secure
image and the bsd library, respectively. Thus we can remove the
compatible property from these nodes. We also make use of
'reserved-memory' to represent the different memory partitions
to be used by the nrf9160 builds.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on boards/arm/nrf5340pdk_nrf5340/Kconfig.defconfig:
```diff
@@ -30,8 +30,8 @@ config PWM_0
 # be restricted to the secure image SRAM partition (sram-secure-partition).
-# Otherwise (if TRUSTED_EXECUTION_SECURE is not set) the whole sram0 may be
-# used by the image.
+# Otherwise (if TRUSTED_EXECUTION_SECURE is not set) the whole zephyr,sram
```
Now it is more generic (as long as we're using `zephyr,sram` for image ram)

galak reviewed (changes requested):
Please use reserved memory as per:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/reserved-memory/reserved-memory.txt

galak reviewed (commented):
Do you need the labels?

galak reviewed (changes requested):
Looks good, outside of the question if you need to label memory nodes (which I'd ask be removed, unless there's a really good reason for it).

ioannisg commented:
> Looks good, outside of the question if you need to label memory nodes (which I'd ask be removed, unless there's a really good reason for it).

No I did not need the labels, that's right. What I wanted is some inline descriptive comments.
fixed now.

galak commented on boards/arm/nrf9160dk_nrf9160/nrf9160dk_nrf9160_common.dts:
```diff
@@ -181,14 +181,19 @@
 / {
+	/* SRAM allocated to the Secure image */
+	sram0_s: memory@20000000 {
+		label = "secure_sram_memory";
```
is label needed here?

galak commented on boards/arm/nrf9160dk_nrf9160/nrf9160dk_nrf9160_common.dts:
```diff
@@ -181,14 +181,19 @@
 	/* SRAM allocated and used by the BSD library */
 	sram0_bsd: memory@20010000 {
-		compatible = "mmio-sram";
+		label = "bsd_shared_memory";
```
is label needed here?

ioannisg replied:
nope, fixed that here as well, thanks

galak reviewed (changes requested):
still a few labels in nrf9160dk_nrf9160_common.dts

mbolivar-nordic commented on boards/arm/nrf5340pdk_nrf5340/nrf5340pdk_nrf5340_cpuapp_common.dts:
```diff
@@ -164,21 +164,23 @@
 		#size-cells = <1>;
+		ranges;
+
+		sram0_image: image@20000000 {
```
Is it OK that this has the same unit address as image_s?

ioannisg replied:
Yes, this is intended  @mbolivar-nordic , this not represents the union of image_s and image_ns

mbolivar-nordic replied:
Doesn't this generate warnings in some dtc / edtlib configurations? Do we have those disabled for our ICs?

ioannisg replied:
No it does not , I guess because the prefix names differ in each node

galak commented on boards/arm/nrf5340pdk_nrf5340/nrf5340pdk_nrf5340_shared_sram_planning_conf.dts:
```diff
@@ -17,10 +17,15 @@
-		device_type = "memory";
-		compatible = "mmio-sram";
-		reg = <0x20070000 DT_SIZE_K(64)>;
+	partitions {
```
Why is this partitions?

galak reviewed (dismissed):
nrf5340pdk_nrf5340_shared_sram_planning_conf.dts should not use partitions.

ioannisg commented:
> nrf5340pdk_nrf5340_shared_sram_planning_conf.dts should not use partitions.

The problem is that I need a "compatible" property to be able to use the new style DTS (see the first commit).

ioannisg commented:
This PR should not be merged, until we have a solution to generate new-style DT macros for nodes without compatible property.

ioannisg committed:
boards: arm: nrf: add non-secure SRAM memory information in DTS

When we build Zephyr as Secure image on nRF340 Application
MCU and nRF9160 SoC we would like to pass the information
about the reserved memory area allocated to the Non-Secure
images. The information may be needed to apply proper
security configuration. We add a "chosen" node in board .dts
file for this purpose.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on boards/arm/nrf5340pdk_nrf5340/nrf5340pdk_nrf5340_shared_sram_planning_conf.dts:
```diff
@@ -8,7 +8,7 @@
  * and nRF5340 CPUNET (Network MCU).
  * - 64 kB SRAM allocated as Shared memory (sram0_shared)
- * - Region defined after the secure SRAM partition of CPUAPP
+ * - Region defined after the image SRAM of Application MCU
```
this is actually a documentation fix

pabigot commented on boards/arm/nrf5340pdk_nrf5340/nrf5340pdk_nrf5340_shared_sram_planning_conf.dts:
```diff
@@ -17,10 +17,14 @@
+		ranges;
+
+		sram0_shared: memory@20070000 {
+		/* SRAM allocated to shared memory */
```
indentation?

ioannisg replied:
done

pabigot commented on boards/arm/nrf5340pdk_nrf5340/nrf5340pdk_nrf5340_cpuapp.dts:
```diff
@@ -14,5 +14,6 @@
 		zephyr,flash = &flash0;
 		zephyr,code-partition = &slot0_partition;
 		zephyr,sram-secure-partition = &sram0_s;
+		zephyr,sram-non-secure-partition = &sram0_ns;
```
There's some (inconsistent) precedent for documenting standard `chosen` properties in `doc/guides/dts/legacy-macros.rst`.  The surrounded text suggests that it should be complete, so these should be added to it, replacing or augmenting the existing `zephyr,sram` entry.

@mbolivar-nordic if this section of that module remains relevant after we remove support for legacy macros perhaps it should move elsewhere before release?

mbolivar-nordic replied:
> @mbolivar-nordic if this section of that module remains relevant after we remove support for legacy macros perhaps it should move elsewhere before release?

That's a good idea. I have added the following to my TODO list for v2.3 DT docs:

- centralize documentation of chosen zephyr nodes to a non-legacy file, provide a reference to them from the intro

If you're OK with letting me take care of that after freeze, I will do so.

ioannisg replied:
@pabigot ^^
(I am ok, let me know if there's anything in need to do here)

ioannisg replied:
@pabigot ^^

pabigot replied:
I'm fine with @mbolivar-nordic doing this later.

ioannisg replied:
thanks @pabigot

pabigot reviewed (commented):
Indentation and possible documentation suggestions.

pabigot commented:
@ioannisg I'm fine with @mbolivar-nordic doing the documentation changes later.  The indentation should be fixed, though.

ioannisg commented:
> @ioannisg I'm fine with @mbolivar-nordic doing the documentation changes later. The indentation should be fixed, though.

Oh, sure, fixed, just don't push now, to not trigger CI

ioannisg commented:
needs #25009

carlescufi reviewed (approved):
Seems to be using the commonly agreed mechanism now.
